src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -173,6 +173,50 @@ HeapWord* PSOldGen::allocate(size_t word_size) {
   return res;
 }
 
+/*
+ * Divide space into blocks, processes block begins at
+ * bottom + block_index  * (IterateBlockSize / HeapWordSize).
+ * NOTE:
+ * - The initial block start address may not be a valid
+ * object address, _start_array is used to correct it.
+ *
+ * - The end address is not necessary to be object address.
+ *
+ * - If there is an object that crosses blocks, it is
+ * processed by the worker that owns the block within
+ * which the object starts.
+ *
+ */
+void PSOldGen::block_iterate(ObjectClosure* cl, uint block_index) {
+  MutableSpace *space = object_space();
+  HeapWord* bottom = space->bottom();
+  HeapWord* top = space->top();
+  size_t block_word_size = IterateBlockSize / HeapWordSize;
+  HeapWord* begin = bottom + block_index * block_word_size;
+
+  assert((block_word_size % (ObjectStartArray::block_size)) == 0,
+         "BLOCK SIZE not a multiple of start_array block");
+
+  // iterate objects in block.
+  HeapWord* end = MIN2(top, begin + block_word_size);
+  // Only iterate if there are objects between begin and end.
+  if (start_array()->object_starts_in_range(begin, end)) {
+    // Process objects in the range, start from finding object at the begining
+    // address. Note that object_start() can return the last object in previous
+    // block, and that object is processed by other worker scanning that block.
+    // So here only focus on objects that fall into the current block.
+    HeapWord* start = start_array()->object_start(begin);
+    if (start < begin) {
+      start += oop(start)->size();
+    }
+    assert(begin <= start && start < end,
+           "object %p must in the range of [%p, %p)\n", start, begin, end);
```

Comment:
use [I]PTR_FORMAT instead of %p.
The assert is too strict too. The addition in line 210 can make start beyond end.